Teuchos: cleanup YAML read-and-broadcast code

ibaned opened the pull request:
The existing implementation suggested that
C was being used for performance, but ignoring
the return value of fread() was causing warnings.
Found a way to use C++ streams that is supposedly
performant and equivalent.

This isn't urgent as I think the old version works, but its cleaner.
Tested this with Albany.

mhoemmen commented on packages/teuchos/comm/src/Teuchos_YamlParameterListHelpers.cpp:
```diff
@@ -39,65 +39,43 @@
+                             std::istreambuf_iterator<char>());
+      int strsize = yamlString.size();
+      broadcast<int, int>(comm, 0, &strsize);
+      broadcast<int, char>(comm, 0, strsize, &yamlString[0]);
```
It's not valid to use `&yamlString[0]` if `yamlString.size() == 0`.  Teuchos does not (yet) require C++11, so you may not use the `data()` method.  The following expression works:
```
yamlString.size() == 0 ? NULL : &yamlString[0]
```

mhoemmen commented on packages/teuchos/comm/src/Teuchos_YamlParameterListHelpers.cpp:
```diff
@@ -39,65 +39,43 @@
+      broadcast<int, int>(comm, 0, &strsize);
+      std::string yamlString;
+      yamlString.resize(strsize);
+      broadcast<int, char>(comm, 0, strsize, &yamlString[0]);
```
See above comment about `&yamlString[0]`.

mhoemmen commented on packages/teuchos/comm/src/Teuchos_YamlParameterListHelpers.cpp:
```diff
@@ -39,65 +39,43 @@
+          std::runtime_error,
+          "Could not open YAML file " << yamlFileName);
+      std::string yamlString((std::istreambuf_iterator<char>(stream)),
+                             std::istreambuf_iterator<char>());
```
Lines 65-66 make me feel vaguely uncomfortable, though they look technically correct.  (The first iterator is `== end()` when it runs out of characters to read, and the second iterator is always `== end()`.)
It could help to have a comment explaining why this works.  Just use my sentence in parentheses.

ibaned replied:
Right, I copied code that was concise but obscure (had to test it to convince myself it works). I've now made them variables, with the latter called `stream_end`, in an attempt to write self-documenting code. Hopefully thats enough.

ibaned commented:
@mhoemmen thanks for the review. I hope 012a8a3 addresses these concerns.

I guess I should do @trilinos/teuchos instead of always assigning Teuchos items to @bartlettroscoe .

mhoemmen commented:
@ibaned wrote:
> I hope 012a8a3 addresses these concerns.

It does; thanks!  I do wish we had more tests for this code.

> I guess I should do @trilinos/teuchos instead of always assigning Teuchos items to @bartlettroscoe .

Ross is both capable and hard working, but alas hasn't figured out how to clone himself yet ;-)  I agree that it's better to tag @trilinos/teuchos and not assign a particular person by default, unless you happen to know that the person is the only one who can do the work.

mhoemmen commented:
btw @ibaned , did you run any Trilinos tests?  Just curious.  Or, does Albany sufficiently exercise these changes?  Either way, I'll be happy to approve and merge in the changes; just want to make sure that Trilinos builds and passes tests with them.

ibaned commented:
Trilinos does build and I tested with Albany. I was planning to run the checkin test and push later. If you think its fine to merge as-is, thats great for me. As you point out, this function isn't really covered in Teuchos tests, so those shouldn't break.

ibaned commented:
> this function isn't really covered in Teuchos tests
nevermind, I was wrong about that:
https://github.com/trilinos/Trilinos/blob/master/packages/teuchos/comm/test/ParameterList/ParameterList_yaml_Parallel.cpp

mhoemmen commented:
@ibaned wrote:
> I was planning to run the checkin test and push later.

If you have push access, I'd rather let you do that, but thanks!

ibaned commented:
tested and pushed as a81e0f4 and 784d491.

mhoemmen commented:
Thanks!!!

bartlettroscoe commented:
NOTE: The default PT build with the checkin-test.py (and checkin-test-sems.sh wrapper) do not currently test with the yamal-cpp TPL enabled.  Therefore, none of this code gets tested (or protected) by the standard CI build.

However, now that SEMS has the yaml-cpp TPL, we should be able to enable it and test it as part of the standard CI build.

ibaned commented:
@bartlettroscoe please let me know how I could do that now and/or let me know when the default PT build is changed to enable it. Thanks !

bartlettroscoe commented:
> please let me know how I could do that now and/or let me know when the default PT build is changed to enable it. 

Let's create a new issue "Add 'yaml-cpp' TPL to Trilinos CI build" and then I will document the steps to add it.  Now that it is provided by the SEMS env, this should be pretty easy.
